Objects/longobject.c

```diff
@@ -4491,38 +4494,65 @@ long_rshift1(PyLongObject *a, Py_ssize_t wordshift, digit remshift)
 {
     PyLongObject *z = NULL;
     Py_ssize_t newsize, hishift, i, j;
-    digit lomask, himask;
+    digit lomask, himask, omitmark;
 
-    if (Py_SIZE(a) < 0) {
-        /* Right shifting negative numbers is harder */
-        PyLongObject *a1, *a2;
-        a1 = (PyLongObject *) long_invert(a);
-        if (a1 == NULL)
-            return NULL;
-        a2 = (PyLongObject *) long_rshift1(a1, wordshift, remshift);
-        Py_DECREF(a1);
-        if (a2 == NULL)
-            return NULL;
-        z = (PyLongObject *) long_invert(a2);
-        Py_DECREF(a2);
+    /* for a positive integrate x
+         -x >> m = -(x >> m)      when the dropped bits are all zeros,
+                 = -(x >> m) -1   otherwise. */
+
+    if (IS_MEDIUM_VALUE(a)) {
+        stwodigits sval;
```

Comment:
I think this whole block could be simpler. If `m = medium_value(a)`, then if `wordshift == 0`, we want `m >> remshift` if `m` is nonnegative, and `~(~m >> remshift)` if `m` is negative. Then there's no need for `Py_ARITHMETIC_RIGHT_SHIFT`, since we're only ever shifting non-negative values.  To convert the result back to a `PyLong`, I'd suggest using `_PyLong_FromSTwoDigits`.

Reply:
To be clear, I'd suggest something along the lines of:
```
    if (IS_MEDIUM_VALUE(a)) {
        stwodigits sval;
        stwodigits x = medium_value(a);
        if (wordshift == 0) {
            sval = x < 0 ? ~(~x >> remshift) : x >> remshift;
        }
        else {
            sval = x < 0 ? -1 : 0;
        }
        return _PyLong_FromSTwoDigits(sval);
    }
```

Reply:
Compiler Explorer is showing that gcc and clang eliminate that `~` branch on x64, but MSVC does not.

https://godbolt.org/z/8hPvGW3WT

Reply:
@sweeneyde: Nice; thanks! I think I'm happy with that - there's a good chance some future version of MSVC will catch up.

Reply:
@sweeneyde: Looks like this spelling fares better on MSVC:
```c
    int shift = wordshift == 0 ? remshift : PyLong_SHIFT;
    stwodigits sval = x < 0 ? ~(~x >> shift) : x >> shift;
```



https://godbolt.org/z/chfKj9E9W